rs/nns/governance/src/proposals/candid_to_self_describing.rs

```diff
@@ -0,0 +1,165 @@
+use crate::pb::v1::{SelfDescribingValue, SelfDescribingValueArray, SelfDescribingValueMap};
+
+use candid::{IDLValue, Int, Nat, types::value::{IDLField, VariantValue}};
+use candid_parser::{IDLArgs, IDLProg, TypeEnv, check_prog};
+use std::{collections::HashMap, str::FromStr};
+
+/// Converts candid arguments to a self-describing value using the candid source schema.
+pub(crate) fn candid_to_self_describing(
+    candid_source: &str,
+    method_name: &str,
+    args: &[u8],
+) -> Result<SelfDescribingValue, String> {
+    let arg = candid_to_idl(candid_source, method_name, args)?;
+    Ok(idl_to_self_describing(arg))
+}
+
+fn candid_to_idl(candid_source: &str, method_name: &str, args: &[u8]) -> Result<IDLValue, String> {
+    // Parse the Candid source
+    let candid_prog = IDLProg::from_str(candid_source)
+        .map_err(|e| format!("Failed to parse candid source: {:?}", e))?;
+
+    let mut type_env = TypeEnv::new();
+    let service = check_prog(&mut type_env, &candid_prog)
+        .map_err(|e| format!("Failed to check candid program: {:?}", e))?
+        .ok_or_else(|| "Failed to parse candid: no service found".to_string())?;
+
+    // Get the method signature
+    let method = type_env
+        .get_method(&service, method_name)
+        .map_err(|e| format!("Failed to get method '{}': {:?}", method_name, e))?;
+
+    // Parse the arguments using the method signature
+    let idl_args = IDLArgs::from_bytes_with_types(args, &type_env, &method.args)
+        .map_err(|e| format!("Failed to parse args: {:?}", e))?;
+
+    if idl_args.args.is_empty() {
+        return Ok(IDLValue::Null);
+    }
+
+    // Check if we have exactly one argument (as expected for NNS functions)
+    if idl_args.args.len() > 1 {
+        return Err(format!(
+            "Expected at most one argument, got {}",
+            idl_args.args.len()
+        ));
+    }
+
+    let arg = idl_args.args.into_iter().next().unwrap();
+    Ok(arg)
+}
+
+fn idl_to_self_describing(idl: IDLValue) -> SelfDescribingValue {
```

Comment:
Maybe

```suggestion
fn candid_value_to_self_describing(candid_value: IDLValue) -> SelfDescribingValue {
```

Even though the type is called `IDL` by the candid library, the value does not represent any piece of an interface definition (language). Rather, it is (decoded) candid value.

Normally, I really try to match existing naming, but the existing naming is just so horrible.